### PR TITLE
Fix typos + add themes and presidio modules

### PR DESCRIPTION
**Pull request by murilommen**
- some general typos across the repo
- themes now exists with the same structure
- refactored input_output types and utility function into their own modules
- themes is unit tested with default configuration file
- presidio_analyzer is added as a module

### CODE REVIEW

**Comment by murilommen** [comment]
@naddeoa could pick your brain here. type subscription on `embeddding_types.py`/`EmbeddingEncoder.encode` is only supported on python 3.9, at least for running the unit tests. what should we do? require > 3.9 to run langkit? or ditch type sub

**Comment by murilommen** [comment]
> @naddeoa could pick your brain here. type subscription on `embeddding_types.py`/`EmbeddingEncoder.encode` is only supported on python 3.9, at least for running the unit tests. what should we do? require > 3.9 to run langkit? or ditch type sub

solved by supporting python 3.8+ for type checks

File: langkit/core/metric.py [hunk 1]
```diff
@@ -108,8 +108,11 @@ class MultiMetric:
     Callable[[], "MetricCreator"],
     Callable[[], List["MetricCreator"]],
     Callable[[], Metric],
+    Callable[[], MultiMetric],
```
Comment (naddeoa) [inline langkit/core/metric.py:111]: These shouldn't be required. The `Metric' is already a union of single and multiple. If it works with single then it should work with multiple

File: langkit/metrics/embeddings_types.py [hunk 2]
```diff
@@ -5,15 +5,15 @@
...
-        ...
-
-
-class TransformerEmbeddingAdapter(EmbeddingEncoder):
+class TransformerEmbeddingAdapter:
     def __init__(self, transformer: SentenceTransformer):
         self._transformer = transformer
 
-    def encode(self, text: List[str]) -> torch.Tensor:
+    def encode(self, text: List[str]) -> "torch.Tensor":
```
Comment (naddeoa) [inline langkit/metrics/embeddings_types.py:12]: Shouldn't be required right?

File: langkit/metrics/embeddings_utils.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import torch
+import torch.nn.functional as F
+
+from langkit.metrics.embeddings_types import EmbeddingEncoder
+
+
+def compute_embedding_similarity(encoder: EmbeddingEncoder, _in: List[str], _out: List[str]) -> torch.Tensor:
+    in_encoded = torch.as_tensor(encoder.encode(_in))
+    out_encoded = torch.as_tensor(encoder.encode(_out))
+    print(f"computing similarities between {in_encoded.shape} and {out_encoded.shape}")
```
Comment (naddeoa) [inline langkit/metrics/embeddings_utils.py:12]: Can remove or make a debug log if its useful

File: langkit/metrics/pii.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    def init() -> None:
+        spacy.load(spacy_model)
+
+    def udf(text: pd.DataFrame) -> MultiMetricResult:
+        entity_types = {
+            "PHONE_NUMBER": f"{input_name}.pii.phone_number",
+            "EMAIL_ADDRESS": f"{input_name}.pii.email_address",
+            "CREDIT_CARD": f"{input_name}.pii.credit_card",
+        }
```
Comment (naddeoa) [inline langkit/metrics/pii.py:38]: We should probably include a lot more of these by default. They're all listed here

https://microsoft.github.io/presidio/supported_entities/

And maybe update pii_presidio_metric to be able to customize them too since there's a lot of country specific ones.

File: langkit/metrics/regexes/regex_loader.py [hunk 5]
```diff
@@ -38,11 +38,11 @@ def load_pattern_string(pattern_file_content: str) -> CompiledPatternGroups:
 
 
 def load_patterns_file(file_path: str) -> CompiledPatternGroups:
-    with open(file_path, "r") as f:
+    with open(file_path, "r", encoding="utf-8") as f:
         return load_pattern_string(f.read())
 
 
-def _is_pattern(obj: Any) -> TypeGuard[Pattern]:
+def _is_pattern(obj: Any) -> bool:
```
Comment (naddeoa) [inline langkit/metrics/regexes/regex_loader.py:45]: we definitely want type guards. Did something break?

File: langkit/metrics/themes/themes.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (naddeoa) [inline langkit/metrics/themes/themes.json:1]: I think we should simplify this module and just convert this into a python dict in a python file, don't worry about any of the json stuff. All of that will be handled in the policy in a single spot.

File: langkit/metrics/themes/themes.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            themes_groups = json.loads(f.read())
+            assert _validate_themes(themes_groups)
+            return cast(Dict[str, List[str]], themes_groups)
+    except FileNotFoundError as e:
+        logger.error(f"Could not find {file_path}")
+        raise e
+    except json.decoder.JSONDecodeError as e:
+        logger.warning(f"Could not parse {file_path}: {e}")
+        raise e
+
```
Comment (naddeoa) [inline langkit/metrics/themes/themes.py:54]: This can hopefully go away.

File: langkit/metrics/themes/themes.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+def __themes_module(column_name: str, themes_group: str, embedding_encoder: Optional[EmbeddingEncoder] = None) -> Metric:
```
Comment (naddeoa) [inline langkit/metrics/themes/themes.py:56]: Lets stick with "metric" over "module" now

File: langkit/metrics/themes/themes.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    # Load the themes into a Dictionary
+    __default_themes_groups: Dict[str, List[str]] = load_themes()
+
+    # Load the encoder
+    if embedding_encoder is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        encoder = TransformerEmbeddingAdapter(SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2", device=device))
+    else:
+        encoder = embedding_encoder
```
Comment (naddeoa) [inline langkit/metrics/themes/themes.py:65]: This stuff should probably also be initializable in an `init` method that is returned in the metric

File: langkit/metrics/themes/themes.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+    def __group_similarity(input_text: pd.DataFrame, input_column_name: str, group: str) -> float:
+        input_to_list: List[str] = UdfInput(text=input_text).to_list(input_column_name)
+
+        similarities = [
+            compute_embedding_similarity(encoder, input_to_list, [themes_strings])
+            for themes_strings in __default_themes_groups.get(group, [])
+        ]
```
Comment (naddeoa) [inline langkit/metrics/themes/themes.py:73]: Isn't this the same as 

```python
theme_samples = __default_themes_groups.get(group, [])
similarities = [compute_embedding_similarity(encoder, input_to_list, theme_samples]
```

No comprehension needed?

File: tests/langkit/metrics/test_pii.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    "ints/min",
+]
+
+
+def _log(item: pd.DataFrame, conf: EvaluationConfig) -> pd.DataFrame:
+    schema = create_whylogs_udf_schema(conf)
+    return why.log(item, schema=schema).view().to_pandas()  # type: ignore
+
+
+def test_prompt_response_themes_module():
```
Comment (naddeoa) [inline tests/langkit/metrics/test_pii.py:48]: nice. Lets add a version that just tests the prompt/response alone too with AAAIII

File: tests/langkit/metrics/test_themes.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    )
+
+    expected_rows = [
+        "prompt",
+        "prompt.jailbreak_similarity",
+        "response",
+        "response.refusal_similarity",
+    ]
+
```
Comment (naddeoa) [inline tests/langkit/metrics/test_themes.py:66]: This actually isn't what I expected. If we're adding the entire module to prompt and response then I would have expected them both to have both, but I guess this is the first module that isn't necessarily symmetric in that way because an LLM can't really jailbreak itself.

But these shouldn't be coming from the "themes" module right? Or should the themes module actually be split into `jailbreak_metric` and `refusal_metric`, with no option to include the prompt/response since they don't make sense? Or maybe themes is just a stand alone thing and we have the `jailbreak_metric` and the `refusal_metric` that build ontop of it, just like the regex file contains the regex metrics as well as the substitution metrics that build on top of the base regex. I think that would be more intuitive since this module is introducing metrics with names that aren't implied at all by the metric itself.

Comment (naddeoa) [reply tests/langkit/metrics/test_themes.py:66]: and then we would expand these tests a little bit to include each version, for each prompt/response

File: langkit/metrics/themes/themes.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        max_similarities = [t.max() for t in similarities]  # type: ignore[reportUnknownMemeberType]
+
+        return torch.max(torch.stack(max_similarities)).item() if max_similarities else 0.0  # type: ignore[reportUnknownMemeberType]
+
+    def udf(text: pd.DataFrame) -> SingleMetricResult:
+        metrics = [__group_similarity(text, column_name, themes_group)]
```
Comment (FelipeAdachi) [inline langkit/metrics/themes/themes.py:79]: Not sure if I'm understanding correctly, but:

This implies that the encoding for both the target promp/response and all samples in the themes group is done at runtime, right?

For a scenario where a workflow is initialized and then `run()` is called multiple times for different dfs, computing the embeddings for the themes group (which should be immutable across runs) at initialization should save a lot of embeddings computing.

Comment (naddeoa) [reply langkit/metrics/themes/themes.py:79]: @FelipeAdachi Yeah you're right, good catch. I'll update this so that the embeddings are computed once as one of the lazy init values.

Comment (naddeoa) [reply langkit/metrics/themes/themes.py]: @FelipeAdachi latest version gets rid of that, and also deletes a bunch of logic in here in favor of vectorized operations with tensors, which was way simpler and faster.
